File: src/perflib.swift
/*! ===========================================================================
@file perflib.swift

@brief A set of utilities used to perform all of the performance measurements.

@copyright 2015 David Owens II. All rights reserved.
============================================================================ */

import Foundation

struct Result {
    let avg: Float
    let min: Float
    let max: Float
    let stddev: Float
}

typealias TimingFunction = () -> ()
typealias TestFunction = (Int, Int) -> Result

// NOTE(owensd): Trying out this faking of namespaces...
enum perflib {
    static var formatter: NSNumberFormatter = {
        let formatter = NSNumberFormatter()
        formatter.paddingCharacter = " "
        formatter.usesGroupingSeparator = false
        formatter.paddingPosition = NSNumberFormatterPadPosition.BeforePrefix
        formatter.numberStyle = NSNumberFormatterStyle.DecimalStyle
        formatter.maximumFractionDigits = 6
        formatter.formatWidth = 8
        
        return formatter
    }()
    
    static func header(description: String) {
        let columnCount = 115
        let timingWidth = 10
        let numberOfTimings = 3
        let stddevWidth = 8
        let titleWidth = columnCount - (numberOfTimings * (timingWidth + 1)) - stddevWidth

        let remainingTitleWidth = titleWidth - description.characters.count
        
        print("\(description)", appendNewline: false)
        for _ in 0 ..< remainingTitleWidth { print(" ", appendNewline: false) }
        print("┃ Avg (ms) ┃ Min (ms) ┃ Max (ms) ┃ StdDev ┃")
        
        for  _ in 0 ..< titleWidth { print("━", appendNewline: false) }
        print("╇", appendNewline: false)
        
        for _ in 0 ..< numberOfTimings {
            for _ in 0 ..< timingWidth { print("━", appendNewline: false) }
            print("╇", appendNewline: false)
        }
        
        for _ in 0 ..< stddevWidth { print("━", appendNewline: false) }
        print("┩")
    }
    
    static func separator(endOfTable: Bool = true) {
        let columnCount = 115
        let timingWidth = 10
        let numberOfTimings = 3
        let stddevWidth = 8
        let titleWidth = columnCount - (numberOfTimings * (timingWidth + 1)) - stddevWidth
        
        for _ in 0 ..< titleWidth { print("─", appendNewline: false) }
        print(endOfTable ? "┴" : "┼", appendNewline: false)
        
        for _ in 0 ..< numberOfTimings {
            for _ in 0 ..< timingWidth { print("─", appendNewline: false) }
            print(endOfTable ? "┴" : "┼", appendNewline: false)
        }
        
        for _ in 0 ..< stddevWidth { print("─", appendNewline: false) }
        print(endOfTable ? "┘" : "┤")
    }

    static func measure(sampleCount: Int, _ iterationCount: Int, _ fn: TimingFunction) -> Result {
        var samples = [Float](count: sampleCount, repeatedValue: 0.0)
        for s in 0 ..< sampleCount {
            var elapsed: UInt64 = 0
            
            for _ in 0 ..< iterationCount {
                let start = mach_absolute_time()
                fn()
                elapsed += mach_absolute_time() - start
            }
            
            samples[s] = Float(elapsed) / Float(NSEC_PER_MSEC)
        }
        
        let avg = samples.reduce(0.0, combine: +) / Float(sampleCount)
        let sums = samples.reduce(0.0) { sum, x in ((x - avg) * (x - avg)) + sum }
        let stddev = sqrt(sums / Float(sampleCount - 1))
        
        let min = samples.minElement() ?? 0.0
        let max = samples.maxElement() ?? 0.0
        
        return Result(avg: avg, min: min, max: max, stddev: stddev)
    }
    
    static func printResult(description: String, _ result: Result) {
        let columnCount = 115
        let timingWidth = 10
        let numberOfTimings = 3
        let stddevWidth = 8
        let titleWidth = columnCount - (numberOfTimings * (timingWidth + 1)) - stddevWidth
        
        let remainingTitleWidth = titleWidth - description.characters.count
        
        print("\(description)", appendNewline: false)
        for _ in 0 ..< remainingTitleWidth { print(" ", appendNewline: false) }
        
        // NSNumberFormatter is a terrible, terrible API...
        func trim(str: String!, len: Int) -> String {
            return str.substringToIndex(advance(str.startIndex, len))
        }
        
        let avg = trim(perflib.formatter.stringFromNumber(result.avg), len: 8)
        let min = trim(perflib.formatter.stringFromNumber(result.min), len: 8)
        let max = trim(perflib.formatter.stringFromNumber(result.max), len: 8)
        let stddev = trim(perflib.formatter.stringFromNumber(result.stddev), len: 6)
        
        print("│ \(avg) │ \(min) │ \(max) │ \(stddev) │")
    }

    static func test(description: String, _ samples: Int, _ iterations: Int, _ fn: TestFunction) {
        let result = fn(samples, iterations)
        printResult(description, result)
    }
}

File: src/perflib.c
/*! ===========================================================================
 @file perflib.c
 
 @brief A set of utilities used to perform all of the performance measurements.
 
 @copyright 2015 David Owens II. All rights reserved.
 ============================================================================ */

typedef struct {
    float Average;
    float Minimum;
    float Maximum;
    float StdDev;
} PLResult;

typedef float (^PLCombiner)(float Accum, float Value);

static inline float
PLReduce(float *Items, int Size, float Initial, PLCombiner Combiner) {
    float Result = Initial;
    for (int Index = 0; Index < Size; ++Index) {
        Result = Combiner(Result, Items[Index]);
    }
    
    return Result;
}

static inline PLResult
PLMeasure(uint32_t SampleCount,
        uint32_t IterationCount,
        void(^FunctionToTime)(void))
{
    int SamplesSize = sizeof(float) * SampleCount;
    float *Samples = malloc(SamplesSize);
    memset(Samples, 0, SamplesSize);
    
    float *Sample = &Samples[0];
    for (int SampleIndex = 0; SampleIndex < SampleCount; ++SampleIndex, ++Sample) {
        uint64_t Elapsed = 0;
        
        for (int IterationIndex = 0; IterationIndex < IterationCount; ++IterationIndex) {
            uint64_t StartTime = mach_absolute_time();
            FunctionToTime();
            Elapsed += mach_absolute_time() - StartTime;
        }
        
        *Sample = (float)Elapsed / NSEC_PER_MSEC;
    }
    
    float Sum = PLReduce(Samples, SampleCount, 0.f, ^(float Accum, float Value) {
        return Accum + Value;
    });
    float Average = Sum / SampleCount;
    
    float Sums = PLReduce(Samples, SampleCount, 0.f, ^(float Accum, float Value) {
        float Diff = Value - Average;
        return Accum + (Diff * Diff);
    });
    float StdDev = sqrtf(Sums / (SampleCount - 1));
    
    float Minimum = 0.f;
    float Maximum = 0.f;
    
    PLResult Result = { Average, Minimum, Maximum, StdDev };
    free(Samples);
    
    return Result;
}

static inline void
PLPrintHeader()
{
    struct winsize TerminalSize;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &TerminalSize);
    
    const int ColumnCount = TerminalSize.ws_col - 5;
    const int TimingWidth = 10;
    const int NumberOfTimings = 3;
    const int StdDevWidth = 8;
    const int TitleWidth = ColumnCount - (NumberOfTimings * (TimingWidth + 1)) - StdDevWidth;
    
    printf("%*s", TitleWidth + TimingWidth + 6, "┃ Avg (ms) ┃");
    printf("%*s", TimingWidth + 3, "Min (ms) ┃");
    printf("%*s", TimingWidth + 3, "Max (ms) ┃");
    printf("%*s\n", StdDevWidth + 3, "StdDev ┃");
    
    for (int ColumnOffset = 0; ColumnOffset < TitleWidth; ++ColumnOffset) {
        printf("━");
    }
    printf("╇");
    
    for (int TimingCount = 0; TimingCount < NumberOfTimings; ++TimingCount) {
        for (int ColumnOffset = 0; ColumnOffset < TimingWidth; ++ColumnOffset) {
            printf("━");
        }
        printf("╇");
    }
    
    for (int ColumnOffset = 0; ColumnOffset < StdDevWidth; ++ColumnOffset) {
        printf("━");
    }
    printf("┩\n");
}

static inline void
PLPrintSeparator(int EndOfTable)
{
    struct winsize TerminalSize;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &TerminalSize);
    
    const int ColumnCount = TerminalSize.ws_col - 5;
    const int TimingWidth = 10;
    const int NumberOfTimings = 3;
    const int StdDevWidth = 8;
    const int TitleWidth = ColumnCount - (NumberOfTimings * (TimingWidth + 1)) - StdDevWidth;
    
    for (int ColumnOffset = 0; ColumnOffset < TitleWidth; ++ColumnOffset) {
        printf("─");
    }
    printf("%s", EndOfTable ? "┴" : "┼");
    
    for (int TimingCount = 0; TimingCount < NumberOfTimings; ++TimingCount) {
        for (int ColumnOffset = 0; ColumnOffset < TimingWidth; ++ColumnOffset) {
            printf("─");
        }
        printf("%s", EndOfTable ? "┴" : "┼");
    }
    
    for (int ColumnOffset = 0; ColumnOffset < StdDevWidth; ++ColumnOffset) {
        printf("─");
    }
    printf("%s\n", EndOfTable ? "┘" : "┤");
}

static inline void
PLPrintResult(const char *Message, PLResult Result) {
    struct winsize TerminalSize;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &TerminalSize);
    
    const int ColumnCount = TerminalSize.ws_col - 5;
    const int TimingWidth = 10;
    const int NumberOfTimings = 3;
    const int StdDevWidth = 8;
    const int TitleWidth = ColumnCount - (NumberOfTimings * (TimingWidth + 1)) - StdDevWidth;

    const int RemainingTitleWidth = TitleWidth - strlen(Message);
    
    printf("%.*s %*s │", TitleWidth, Message, RemainingTitleWidth - 2, " ");
    printf("%*.3f │", TimingWidth - 1, Result.Average);
    printf("%*.3f │", TimingWidth - 1, Result.Minimum);
    printf("%*.3f │", TimingWidth - 1, Result.Maximum);
    printf("%*.3f │\n", StdDevWidth - 1, Result.StdDev);
}


